core/services/ocr2/plugins/s4/plugin.go

```diff
@@ -0,0 +1,288 @@
+}
+
+var _ types.ReportingPlugin = (*plugin)(nil)
+
+func NewConsensusPlugin(config *PluginConfig, orm s4.ORM) (*plugin, error) {
+	if config.NSnapshotShards == 0 {
+		return nil, errors.New("number of snapshots shards cannot be zero")
+	}
+	if config.MaxObservationEntries == 0 {
+		return nil, errors.New("max number of observation entries cannot be zero")
+	}
+	if config.MaxReportEntries == 0 {
+		return nil, errors.New("max number of report entries cannot be zero")
+	}
+
+	addressRange := new(big.Int).Sub(s4.MaxAddress, s4.MinAddress)
+	divisor := big.NewInt(int64(config.NSnapshotShards))
+	addressInteval := new(big.Int).Div(addressRange, divisor)
+
+	return &plugin{
+		config:          config,
+		orm:             orm,
+		minQueryAddress: s4.MinAddress,
+		maxQueryAddress: new(big.Int).Add(s4.MinAddress, addressInteval),
+		addressInteval:  addressInteval,
+	}, nil
+}
+
+func (c *plugin) advanceAddressRange() {
+	if c.config.NSnapshotShards > 1 {
+		c.minQueryAddress = new(big.Int).Add(c.minQueryAddress, c.addressInteval)
+		c.maxQueryAddress = new(big.Int).Add(c.maxQueryAddress, c.addressInteval)
+		if c.maxQueryAddress.Cmp(s4.MaxAddress) > 0 {
+			c.maxQueryAddress = s4.MaxAddress
+		}
+		if c.minQueryAddress.Cmp(s4.MaxAddress) >= 0 {
+			c.minQueryAddress = s4.MinAddress
+			c.maxQueryAddress = new(big.Int).Add(c.minQueryAddress, c.addressInteval)
+		}
+	}
+}
+
+func (c *plugin) Query(ctx context.Context, _ types.ReportTimestamp) (types.Query, error) {
+	snapshot, err := c.orm.GetSnapshot(c.minQueryAddress, c.maxQueryAddress, pg.WithParentCtx(ctx))
+	if err != nil {
+		return nil, err
+	}
+
+	if len(snapshot) == 0 {
+		return nil, nil
+	}
+
+	rows := make([]*Row, len(snapshot))
+	for i, row := range snapshot {
+		rows[i] = convertRow(row)
+	}
+	query, err := marshalRows(rows, c.minQueryAddress, c.maxQueryAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	c.advanceAddressRange()
+
+	return query, err
+}
+
+func (c *plugin) Observation(ctx context.Context, _ types.ReportTimestamp, query types.Query) (types.Observation, error) {
+	if err := c.orm.DeleteExpired(pg.WithParentCtx(ctx)); err != nil {
+		return nil, err
+	}
+
+	queryRows, minAddress, maxAddress, err := unmarshalRows(query)
+	if err != nil {
+		return nil, err
+	}
+
+	snapshot, err := c.orm.GetSnapshot(minAddress, maxAddress, pg.WithParentCtx(ctx))
+	if err != nil {
+		return nil, err
+	}
+
+	snapshotMap := make(map[key]*Row)
+	unconfirmedMap := make(map[key]*Row)
+	for _, row := range snapshot {
+		r := convertRow(row)
+		mkey := key{
+			address: r.Address,
+			slot:    uint(r.Slotid),
+		}
+		snapshotMap[mkey] = r
+		if !row.Confirmed {
+			unconfirmedMap[mkey] = r
+		}
+	}
+
+	observation := make([]*Row, 0)
+	for _, queryRow := range queryRows {
+		mkey := key{
+			address: queryRow.Address,
+			slot:    uint(queryRow.Slotid),
+		}
+		snapshotRow, ok := snapshotMap[mkey]
+		if ok && queryRow.Version < snapshotRow.Version {
+			observation = append(observation, queryRow)
```

Comment:
You may have NEW in both places having different versions. Why prioritize NEW(v1) over CONFIRMED(v2)? IMO versions have higher precedence.

Reply:
What I'm saying is that you go over the Query first. A malicious leader can pretend they have empty state and force others to send pointless data. We don't want this behavior to starve legitimate NEW entries from users that need to be propagated to honest nodes. That's why I proposed to always prioritize local NEW entries over what was requested in the query.

Reply:
> A malicious leader can pretend they have empty state and force others to send pointless data

Not pointless. In case of empty query, Observation will send all NEW entries for the given range.
I still need your opinion on why we need to prefer NEW(v1) over CONFIRMED(v2).
In addition, I feel that we need to verify signatures in Observation, because we shall not trust Query data.

Reply:
I've added intervals comparison and signature validation here to minimize impact from a malicious leader.